### ISSUE TITLE
Bug: instacrash of `fluid.knnclassifier~` when loading empty dictionary

### ISSUE THREAD

**Issue by rconstanzo** [flucoma/flucoma-core#274]
### Please tell us what you were doing! You can include code and files by drag and dropping them into the text area.

I found an instacrash when you try and `load` a `dict` into `fluid.knnclassifier~` that was created by a `fluid.classifier~` *prior* to having fed it any data.

When you `dump` or `write` a `fluid.knnclassifier~` without data empty you get the following dictionary contents:
```
{
	"labels" : 	{
		"cols" : 1
	}
,
	"tree" : 	{
		"cols" : 0,
		"data" : null,
		"ids" : [  ],
		"rows" : 0,
		"tree" : null
	}

}
```

Here's a patch to reproduce:
![Screenshot 2024-05-20 at 12 01 56 AM](https://github.com/flucoma/flucoma-core/assets/2747015/1dece0a6-d5e3-4ec8-ada3-8463f8703403)

```
<pre><code>
----------begin_max5_patcher----------
723.3ocwV0zbZCCD8r8uBUclxfr4iPO0i8R+EzoCivdgnTaIOxxDRxj9auRq
rHl.NAHz1CVy30qzaeO8zZ8TbDcoZKTSIeg7CRTzSwQQXHWfn12ink7sYE7Z
LMZlprDjF5.+2LvVCFOYHISyquk7c91OE9ZgPBYpFIlRRavUJooV7H3hwlNb
TmvRdIFl9MnXCXDY7vJIaJExBvfEAqMXE2jcqPtdgFxLdRjlZWPxTlab1Mtw
T6H4murLpFSXc5B8JdFzcsE4XgnVd2mmQcgdNN1ML3CpSrgjMB3dRtHyPxr.
aSp9+mdwtAErjYtQFK85nXoWhhsrwXTxyk.SvRlMZLRici8T+gExGx7PE3WE
5RtbMc2j5PkozcXqspsAzK.IeYANwQW.Mkv8108.ewiEDMr9nrO4M19Rlize
xrvFmk9I8R+j9nOcf84XzmcI6jNu8PmK+b2LS7GbYiQ5jjN0StouqerSEO4e
m2iM12g4uk2a7+HuG1KR1TTP9JTtDxIry1G5Uh4I91Hr2wGNoOovUIBkjqe.
cjceNl.kjFJzbtg2R7Vla6lxWBE0gngvQ1Vy9nLe.q70NAiFf9Su0q0ALmjs
KnHu8+ngB0FSqt+USM.gapA3i6VEzZ9FHegkcVwcA2XzBq4DpeM4vMpN1JKc
g07rG5B2w8MG9wRdU0d1p1h553tVUzHxG9KoDyTrR.5ee1mz784RF6a2MZ10
uQ2b50ixUZnBj4jBEO+Ro5H7rzr4uCS6sqROrbuF5XJ3sNd0c.w5xEeetWqZ
zYA.Z+6H4kJKGpMBI2cBtSNLeNGUcOUbFeB33ZF7gAJrHuIRS1KGkNGz82Q9
phrSKYGhL6CgryTbBPO+JHumFPG5X7FUaapMfttMcDD6ww6TnFLe.9pP5eEa
yQ0vFQHeeDt1dfyXOs0n8Wcd6T+uYokJqZJaDsBJ1PlhG0cWyttp85tXGg3m
i+ymqi.X
-----------end_max5_patcher-----------
</code></pre>
```

### What was the expected result?

It should either ignore the loaded `dict` or throw an error message.

### What was the actual result?

Instacrash!

### What operating system were you using?

Mac

### Operating system version

macOS 14.4.1

### FluCoMa Version

1.0.7

**Comment by tremblap** [comment]
confirmed in SC with the following code:
```
d = Dictionary.newFrom(["labels",Dictionary.newFrom(["cols",1]), "tree",Dictionary.newFrom(["cols",0, "data", "null", "ids", [], "rows", 0, "tree", "null" ]),])
b = FluidKNNClassifier(s).load(d)
```

**Comment by tremblap** [comment]
ok, I have a question for @weefuzzy here. We know that it shouldn't crash, but @rconstanzo is sending a corrupted json in - it is produced by us when, as he stated, he saves an empty KNNClassifier.

Let's make a detour via MLPClassifier. If one does the same process (save an empty MLPClassifier and try to load it) one gets an error (invalid JSON format) which is good. no crash.

Now, the c++ is a bit beyond me, but how could we modify KNNClassifier to have a default 'invalid'. Or maybe the solution is in the check that is not happening? I am a bit lost here, help is welcome.

**Comment by weefuzzy** [comment]
ISTM that several things shouldn't happen here, starting with we shouldn't emit invalid JSON in the first place and then, yes, shouldn't crash when it's encountered. 

Preusmably the problem is that the `kdtree` `check_json` isn't returning an error when it ought to 

https://github.com/flucoma/flucoma-core/blob/d44b5ea7043d6313100b81aefe85c27f7c89480c/include/data/FluidJSON.hpp#L147-L154

so it says aok, then tries to `unflatten` and kaboom.

**Comment by weefuzzy** [comment]
So we need to: 
definitely (1) handle that 'empty tree' JSON gracefully (i.e. just make an empty kdtree from it)
perhaps (2)  refuse to emit such a thing.

**Comment by rconstanzo** [comment]
Without knowing the internals, I would imagine nothing that has some kind of state should save/write before it has one. I personally never tried, but someone using SP-Tools hit `save` before finishing a train loop, which led to this discovery.

Then ideally the thing would be robust to loading any garbage `.json` (e.g. accidentally loading another kind of fit/dataset shouldn't crash Max by mistake).

**Comment by tremblap** [comment]
So, @weefuzzy, to be rigorous here (as you know me). Should I try for all our saving object, to save empty and load and see what happens. Then make a report, and find a consistent solution, since now we have KNNClass and MLPClass not doing the same thing...

**Comment by weefuzzy** [comment]
Not doing the same thing as in one crashes and the other doesn't, or some different divergence? 

Anyway, the important bit is not crashing, so yes, please check the other data objects!

**Comment by tremblap** [comment]
fixed in #282 282
